File: sdk/python/gitclaw/async_clients/trending.py
"""Async Trending resource client.

Design Reference: DR-5
Requirements: 11.1, 11.2, 11.3, 1.6
"""

from datetime import datetime
from typing import TYPE_CHECKING

from gitclaw.types.trending import TrendingRepo, TrendingResponse

if TYPE_CHECKING:
    from gitclaw.async_transport import AsyncHTTPTransport


class AsyncTrendingClient:
    """Async client for trending repository discovery."""

    def __init__(self, transport: "AsyncHTTPTransport") -> None:
        """
        Initialize the async trending client.

        Args:
            transport: Async HTTP transport for making requests
        """
        self.transport = transport

    async def get(
        self,
        window: str = "24h",
        limit: int = 50,
    ) -> TrendingResponse:
        """
        Get trending repositories.

        This is an unsigned request - no authentication required.
        Results are sorted by weighted_score in descending order.

        Args:
            window: Time window for trending calculation
                    ("1h", "24h", "7d", "30d", default: "24h")
            limit: Maximum number of results (1-100, default: 50)

        Returns:
            TrendingResponse with repos sorted by weighted_score
        """
        params: dict[str, str | int] = {
            "window": window,
            "limit": limit,
        }

        response = await self.transport.unsigned_request(
            method="GET",
            path="/v1/repos/trending",
            params=params,
        )

        data = response.get("data", {})
        repos = [
            TrendingRepo(
                repo_id=repo["repoId"],
                name=repo["name"],
                owner_id=repo["ownerId"],
                owner_name=repo["ownerName"],
                description=repo.get("description"),
                stars=repo["stars"],
                stars_delta=repo["starsDelta"],
                weighted_score=repo["weightedScore"],
                created_at=datetime.fromisoformat(repo["createdAt"].rstrip("Z")),
            )
            for repo in data.get("repos", [])
        ]

        return TrendingResponse(
            window=data["window"],
            repos=repos,
            computed_at=datetime.fromisoformat(data["computedAt"].rstrip("Z")),
        )

File: backend/src/models/trending.rs
//! Trending model and related types
//!
//! Models for trending repositories functionality.
//! Design Reference: DR-12.1 (Trending Service)

use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use std::fmt;
use std::str::FromStr;

/// Trending time window
#[derive(Debug, Clone, Copy, Default, PartialEq, Eq, Serialize, Deserialize, sqlx::Type)]
#[sqlx(type_name = "trending_window", rename_all = "lowercase")]
pub enum TrendingWindow {
    #[serde(rename = "1h")]
    #[sqlx(rename = "1h")]
    OneHour,
    #[serde(rename = "24h")]
    #[sqlx(rename = "24h")]
    #[default]
    TwentyFourHours,
    #[serde(rename = "7d")]
    #[sqlx(rename = "7d")]
    SevenDays,
    #[serde(rename = "30d")]
    #[sqlx(rename = "30d")]
    ThirtyDays,
}

impl fmt::Display for TrendingWindow {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Self::OneHour => write!(f, "1h"),
            Self::TwentyFourHours => write!(f, "24h"),
            Self::SevenDays => write!(f, "7d"),
            Self::ThirtyDays => write!(f, "30d"),
        }
    }
}

impl FromStr for TrendingWindow {
    type Err = String;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s {
            "1h" => Ok(Self::OneHour),
            "24h" => Ok(Self::TwentyFourHours),
            "7d" => Ok(Self::SevenDays),
            "30d" => Ok(Self::ThirtyDays),
            _ => Err(format!(
                "Invalid window: {s}. Valid values are: 1h, 24h, 7d, 30d"
            )),
        }
    }
}

impl TrendingWindow {
    /// Get the duration in hours for this window
    pub fn hours(&self) -> i64 {
        match self {
            Self::OneHour => 1,
            Self::TwentyFourHours => 24,
            Self::SevenDays => 24 * 7,
            Self::ThirtyDays => 24 * 30,
        }
    }
}

/// Query parameters for trending endpoint
#[derive(Debug, Clone, Deserialize)]
pub struct TrendingQuery {
    /// Time window for trending calculation (default: 24h)
    #[serde(default)]
    pub window: Option<String>,
    /// Maximum number of results to return (default: 50)
    #[serde(default)]
    pub limit: Option<i32>,
}

/// Trending repository information
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TrendingRepo {
    pub repo_id: String,
    pub name: String,
    pub owner_id: String,
    pub owner_name: String,
    pub description: Option<String>,
    pub stars: i32,
    pub stars_delta: i32,
    pub weighted_score: f64,
    pub created_at: DateTime<Utc>,
}

/// Response for trending repositories endpoint
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TrendingResponse {
    pub window: String,
    pub repos: Vec<TrendingRepo>,
    pub computed_at: Option<DateTime<Utc>>,
}
